light-clients/ics07-tendermint/src/client_state.rs

```diff
@@ -0,0 +1,677 @@
+				"ClientState latest height must be greater than zero".to_string(),
+			))
+		}
+
+		// `TrustThreshold` is guaranteed to be in the range `[0, 1)`, but a `TrustThreshold::ZERO`
+		// value is invalid in this context
+		if trust_level == TrustThreshold::ZERO {
+			return Err(Error::validation("ClientState trust-level cannot be zero".to_string()))
+		}
+
+		// Disallow empty proof-specs
+		if proof_specs.is_empty() {
+			return Err(Error::validation("ClientState proof-specs cannot be empty".to_string()))
+		}
+
+		Ok(Self {
+			chain_id,
+			trust_level,
+			trusting_period,
+			unbonding_period,
+			max_clock_drift,
+			latest_height,
+			proof_specs,
+			upgrade_path,
+			frozen_height: None,
+			_phantom: Default::default(),
+		})
+	}
+
+	pub fn latest_height(&self) -> Height {
+		self.latest_height
+	}
+
+	pub fn with_header(self, h: Header) -> Self {
+		// TODO: Clarify which fields should update.
+		ClientState {
+			latest_height: self
+				.latest_height
+				.with_revision_height(u64::from(h.signed_header.header.height)),
+			..self
+		}
+	}
+
+	pub fn with_frozen_height(self, h: Height) -> Result<Self, Error> {
+		if h == Height::zero() {
+			return Err(Error::validation(
+				"ClientState frozen height must be greater than zero".to_string(),
+			))
+		}
+		Ok(Self { frozen_height: Some(h), ..self })
+	}
+
+	/// Get the refresh time to ensure the state does not expire
+	pub fn refresh_time(&self) -> Option<Duration> {
+		Some(2 * self.trusting_period / 3)
+	}
+
+	pub fn chain_id(&self) -> ChainId {
+		self.chain_id.clone()
+	}
+
+	pub fn client_type() -> ClientType {
+		"07-tendermint"
+	}
+
+	pub fn frozen_height(&self) -> Option<Height> {
+		self.frozen_height
+	}
+
+	pub fn upgrade(
+		mut self,
+		upgrade_height: Height,
+		upgrade_options: UpgradeOptions,
+		chain_id: ChainId,
+	) -> Self {
+		// Reset custom fields to zero values
+		self.reset();
+
+		// Upgrade the client state
+		self.latest_height = upgrade_height;
+		self.unbonding_period = upgrade_options.unbonding_period;
+		self.chain_id = chain_id;
+
+		self
+	}
+
+	fn reset(&mut self) {
+		self.trusting_period = ZERO_DURATION;
+		self.trust_level = TrustThreshold::ZERO;
+		self.frozen_height = None;
+		self.max_clock_drift = ZERO_DURATION;
+	}
+
+	/// Check if the state is expired when `elapsed` time has passed since the latest consensus
+	/// state timestamp
+	pub fn expired(&self, elapsed: Duration) -> bool {
+		elapsed > self.trusting_period
+	}
+
+	/// Helper method to produce a [`Options`] struct for use in
+	/// Tendermint-specific light client verification.
+	pub fn as_light_client_options(&self) -> Result<Options, Error> {
+		Ok(Options {
+			trust_threshold: self
+				.trust_level
+				.try_into()
+				.map_err(|e: Ics02Error| Error::invalid_trust_threshold(e.to_string()))?,
+			trusting_period: self.trusting_period,
+			clock_drift: self.max_clock_drift,
+		})
+	}
+
+	/// Verify the time and height delays
+	pub fn verify_delay_passed(
```

Comment:
we can move this function to the `common` lib, it's implementation would be the same for all the light clients.